File: android_mobile/src/main/java/com/alexstyl/specialdates/contact/AndroidContactFactory.java
package com.alexstyl.specialdates.contact;

import android.content.ContentResolver;
import android.content.ContentUris;
import android.database.Cursor;
import android.net.Uri;
import android.provider.ContactsContract;
import android.provider.ContactsContract.Contacts;
import android.support.annotation.NonNull;

import com.alexstyl.specialdates.ErrorTracker;

import java.net.URI;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static com.alexstyl.specialdates.contact.AndroidContactsQuery.SORT_ORDER;

class AndroidContactFactory {

    private static final String WHERE = ContactsContract.Data.IN_VISIBLE_GROUP + "=1";
    private static final String SELECTION_CONTACT_WITH_ID = Contacts._ID + " = ?";


    private final ContentResolver resolver;

    AndroidContactFactory(ContentResolver contentResolver) {
        resolver = contentResolver;
    }

    List<Contact> getAllContacts() {
        Cursor cursor = resolver.query(
                AndroidContactsQuery.CONTENT_URI,
                AndroidContactsQuery.PROJECTION,
                WHERE,
                null,
                AndroidContactsQuery.SORT_ORDER
        );
        List<Contact> contacts = new ArrayList<>();
        try {
            while (cursor.moveToNext()) {
                Contact contact = createContactFrom(cursor);
                contacts.add(contact);
            }
        } catch (Exception e) {
            ErrorTracker.track(e);
        } finally {
            cursor.close();
        }
        return Collections.unmodifiableList(contacts);
    }

    Contact createContactWithId(long contactID) throws ContactNotFoundException {
        Cursor cursor = queryContactsWithContactId(contactID);
        if (isInvalid(cursor)) {
            throw new RuntimeException("Cursor was invalid");
        }
        try {
            if (cursor.moveToFirst()) {
                return createContactFrom(cursor);
            }
        } finally {
            cursor.close();
        }
        throw new ContactNotFoundException(contactID);
    }

    private Contact createContactFrom(Cursor cursor) {
        long contactID = getContactIdFrom(cursor);
        DisplayName displayName = getDisplayNameFrom(cursor);
        Uri lookupKey = getLookupKeyFrom(cursor);
        URI imagePath = URI.create(ContentUris.withAppendedId(Contacts.CONTENT_URI, contactID).toString());
        return new AndroidContact(contactID, displayName, imagePath, lookupKey);
    }

    private Cursor queryContactsWithContactId(long contactID) {
        return resolver.query(
                AndroidContactsQuery.CONTENT_URI,
                AndroidContactsQuery.PROJECTION,
                SELECTION_CONTACT_WITH_ID,
                makeSelectionArgumentsFor(contactID),
                SORT_ORDER + " LIMIT 1"
        );
    }

    @NonNull
    private String[] makeSelectionArgumentsFor(long contactID) {
        return new String[]{
                String.valueOf(contactID)
        };
    }

    private static boolean isInvalid(Cursor cursor) {
        return cursor == null || cursor.isClosed();
    }

    private static long getContactIdFrom(Cursor cursor) {
        return cursor.getLong(AndroidContactsQuery.CONTACT_ID);
    }

    private static DisplayName getDisplayNameFrom(Cursor cursor) {
        return DisplayName.from(cursor.getString(AndroidContactsQuery.DISPLAY_NAME));
    }

    private static Uri getLookupKeyFrom(Cursor cursor) {
        return Uri.parse(cursor.getString(AndroidContactsQuery.LOOKUP_KEY));
    }

}

File: android_mobile/src/main/java/com/alexstyl/specialdates/contact/AndroidContactFactory.kt
package com.alexstyl.specialdates.contact

import android.annotation.SuppressLint
import android.content.ContentResolver
import android.content.ContentUris
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.provider.ContactsContract
import com.alexstyl.specialdates.contact.AndroidContactsQuery.SORT_ORDER
import com.alexstyl.specialdates.contact.ContactSource.SOURCE_DEVICE
import java.net.URI

class AndroidContactFactory(private val resolver: ContentResolver) {

    fun getAllContacts(): List<Contact> {
        val cursor: Cursor = resolver.safeQuery(
                AndroidContactsQuery.CONTENT_URI,
                AndroidContactsQuery.PROJECTION,
                WHERE,
                null,
                SORT_ORDER
        )

        return cursor.use {
            return@use List(it.count) { index ->
                it.moveToPosition(index)
                createContactFrom(it)
            }
        }
    }

    @Throws(ContactNotFoundException::class)
    fun createContactWithId(contactID: Long): Contact {
        queryContactsWithContactId(contactID).use {
            if (it.moveToFirst()) {
                return createContactFrom(it)
            }
        }
        throw ContactNotFoundException(contactID)
    }


    fun queryContacts(ids: List<Long>): List<Contact> {
        val cursor = queryContactsWithContactId(ids)

        return cursor.use {
            return@use List(it.count) { index ->
                it.moveToPosition(index)
                createContactFrom(it)
            }
        }
    }

    private fun queryContactsWithContactId(ids: List<Long>): Cursor {
        return resolver.safeQuery(
                AndroidContactsQuery.CONTENT_URI,
                AndroidContactsQuery.PROJECTION,
                "${AndroidContactsQuery._ID} IN (${ids.joinToString(",")})",
                null,
                SORT_ORDER
        )
    }


    private fun createContactFrom(cursor: Cursor): Contact {
        val contactID = getContactIdFrom(cursor)
        val displayName = getDisplayNameFrom(cursor)
        val imagePath = ContentUris.withAppendedId(AndroidContactsQuery.CONTENT_URI, contactID).toString()
        return Contact(contactID, displayName, imagePath, SOURCE_DEVICE)
    }

    private fun queryContactsWithContactId(contactID: Long): Cursor {
        return resolver.safeQuery(
                AndroidContactsQuery.CONTENT_URI,
                AndroidContactsQuery.PROJECTION,
                SELECTION_CONTACT_WITH_ID,
                makeSelectionArgumentsFor(contactID),
                "$SORT_ORDER LIMIT 1"
        )
    }

    private fun makeSelectionArgumentsFor(contactID: Long): Array<String> =
            arrayOf(contactID.toString())

    companion object {
        private const val WHERE = ContactsContract.Data.IN_VISIBLE_GROUP + " = 1"
        private const val SELECTION_CONTACT_WITH_ID = AndroidContactsQuery._ID + " = ?"
        private val emptyURI = URI.create("")

        private fun getContactIdFrom(cursor: Cursor): Long =
                cursor.getLong(AndroidContactsQuery.CONTACT_ID)


        private fun getDisplayNameFrom(cursor: Cursor): DisplayName {
            val string = cursor.getString(AndroidContactsQuery.DISPLAY_NAME)
            return DisplayName.from(string)
        }
    }

}

@SuppressLint("Recycle")
private fun ContentResolver.safeQuery(uri: Uri,
                                      projection: Array<String>?,
                                      selection: String?,
                                      selectionArgs: Array<String>?,
                                      sortOrder: String?): Cursor =
        this.query(uri, projection, selection, selectionArgs, sortOrder) ?: MatrixCursor(projection)

